bigquerydb/client.go

```diff
@@ -180,3 +184,155 @@ func (c *BigqueryClient) Describe(ch chan<- *prometheus.Desc) {
+			}
+			continue
+		}
+
+		// Labels
+		switch m.Type {
+		case prompb.LabelMatcher_EQ:
+			matchers = append(matchers, fmt.Sprintf("JSON_EXTRACT(tags, '$.%q') = '%s')", m.Name, escapeSingleQuotes(m.Value)))
+		case prompb.LabelMatcher_NEQ:
+			matchers = append(matchers, fmt.Sprintf("JSON_EXTRACT(tags, '$.%q') = '%s')", m.Name, escapeSingleQuotes(m.Value)))
+		case prompb.LabelMatcher_RE:
+			matchers = append(matchers, fmt.Sprintf("REGEXP_CONTAINS(JSON_EXTRACT(tags, '$.%q'), r'%s')", m.Name, escapeSlashes(m.Value)))
+		case prompb.LabelMatcher_NRE:
+			matchers = append(matchers, fmt.Sprintf("not REGEXP_CONTAINS(JSON_EXTRACT(tags, '$.%q'), r'%s')", m.Name, escapeSlashes(m.Value)))
+		default:
+			return "", errors.Errorf("unknown match type %v", m.Type)
+		}
+	}
+	matchers = append(matchers, fmt.Sprintf("timestamp >= TIMESTAMP_MILLIS(%v)", q.StartTimestampMs))
+	matchers = append(matchers, fmt.Sprintf("timestamp <= TIMESTAMP_MILLIS(%v)", q.EndTimestampMs))
+
+	query := fmt.Sprintf("SELECT metricname, tags, timestamp, value FROM %s.%s WHERE %v", c.datasetID, c.tableID, strings.Join(matchers, " AND "))
+	level.Debug(c.logger).Log("msg", "BiQuery read", "sql query", query)
+
+	return query, nil
+}
+
+// rowsToTimeseries iterates over the BigQuery data and creates time series for Prometheus
+func rowsToTimeseries(iter *bigquery.RowIterator) ([]*prompb.TimeSeries, error) {
+	if iter == nil {
+		return nil, nil
+	}
+	tsMap := make(map[model.Fingerprint]*prompb.TimeSeries)
+	for {
+		row := make(map[string]bigquery.Value)
+		err := iter.Next(&row)
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		sample, metric, labels, err := rowToSample(row)
+		if err != nil {
+			return nil, err
+		}
+		fp := metric.Fingerprint()
+		if _, ok := tsMap[fp]; !ok {
+			tsMap[fp] = &prompb.TimeSeries{Labels: labels}
+		}
+		tsMap[fp].Samples = append(tsMap[fp].Samples, *sample)
+	}
+
+	return orderTimeSeries(tsMap), nil
+}
+
+// rowToSample converts a BigQuery row to a sample and also processes the labels for later consumption
+func rowToSample(row map[string]bigquery.Value) (*prompb.Sample, model.Metric, []*prompb.Label, error) {
+	var v interface{}
+	labelsJSON := row["tags"].(string)
+	json.Unmarshal([]byte(labelsJSON), &v)
+	labels := v.(map[string]interface{})
+	labelPairs := make([]*prompb.Label, 0, len(labels))
+	metric := model.Metric{}
+	for name, value := range labels {
+		labelPairs = append(labelPairs, &prompb.Label{
+			Name:  name,
+			Value: value.(string),
+		})
+		metric[model.LabelName(name)] = model.LabelValue(value.(string))
+	}
+	labelPairs = append(labelPairs, &prompb.Label{
+		Name:  model.MetricNameLabel,
+		Value: row["metricname"].(string),
+	})
+	metric[model.LabelName(model.MetricNameLabel)] = model.LabelValue(row["metricname"].(string))
+	return &prompb.Sample{Timestamp: row["timestamp"].(time.Time).Unix(), Value: row["value"].(float64)}, metric, labelPairs, nil
```

Comment:
Since this is the only `return` statement in this function do you think it makes sense to return an always `nil` error?

Reply:
The error could just be remvoved from the function signature.